subsys/zigbee/osif/zb_nrf_async_serial.c

```diff
@@ -0,0 +1,405 @@
+	k_sem_give(&tx_done_sem);
+
+	if (tx_trx_data_cb) {
+		zigbee_schedule_callback(
+			tx_trx_data_cb,
+			SERIAL_SEND_TIMEOUT_EXPIRED);
+		tx_trx_data_cb = NULL;
+	}
+}
+
+/**
+ * Inform user about reception timeout.
+ */
+static void uart_rx_timeout(struct k_timer *dummy)
+{
+	if (uart_rx_buf) {
+		uart_rx_buf_len = 0;
+
+		if (zigbee_schedule_callback(uart_rx_notify, 0)) {
+			uart_rx_buf_offset = 0;
+			uart_rx_buf = NULL;
+			k_sem_give(&rx_done_sem);
+		}
+	}
+}
+
+static void handle_rx_ready_evt(const struct device *dev)
+{
+	int recv_len = 0;
+	uint8_t buffer[CONFIG_ZIGBEE_UART_RX_BUF_LEN] = {0};
+
+	/* Copy data to the user's buffer. */
+	if (uart_rx_buf && (uart_rx_buf_offset < uart_rx_buf_len)) {
+		uart_rx_buf_offset += ring_buf_get(
+			&rx_ringbuf,
+			&uart_rx_buf[uart_rx_buf_offset],
+			uart_rx_buf_len - uart_rx_buf_offset);
+
+		recv_len = uart_fifo_read(
+			dev,
+			&uart_rx_buf[uart_rx_buf_offset],
+			uart_rx_buf_len - uart_rx_buf_offset);
+
+		uart_rx_bytes(&uart_rx_buf[uart_rx_buf_offset], recv_len);
+		uart_rx_buf_offset += recv_len;
+
+		if (uart_rx_buf_offset == uart_rx_buf_len) {
+			uart_rx_buf_len = 0;
+			k_timer_stop(&uart_rx_timer);
+
+			if (zigbee_schedule_callback(uart_rx_notify, 0)) {
+				uart_rx_buf_offset = 0;
+				uart_rx_buf = NULL;
+				k_sem_give(&rx_done_sem);
+			}
+		} else {
+			k_timer_start(
+				&uart_rx_timer,
+				K_MSEC(CONFIG_ZIGBEE_UART_PARTIAL_RX_TIMEOUT),
+				K_NO_WAIT);
+		}
+	} else {
+		recv_len = uart_fifo_read(dev, buffer, sizeof(buffer));
+		uart_rx_bytes(buffer, recv_len);
+
+		/* Store remaining bytes inside the ring buffer. */
+		if (recv_len > ring_buf_space_get(&rx_ringbuf)) {
+			uint8_t dummy_buffer[CONFIG_ZIGBEE_UART_RX_BUF_LEN];
+
+			(void)ring_buf_get(&rx_ringbuf, dummy_buffer,
+				recv_len - ring_buf_space_get(&rx_ringbuf));
+		}
+
+		(void)ring_buf_put(&rx_ringbuf, buffer, recv_len);
+	}
+}
+
+static void handle_tx_ready_evt(const struct device *dev)
+{
+	if (uart_tx_buf_len <= uart_tx_buf_offset) {
+		uart_irq_tx_disable(dev);
+		return;
+	}
+
+	uart_tx_buf_offset += uart_fifo_fill(
+		dev,
+		&uart_tx_buf[uart_tx_buf_offset],
+		uart_tx_buf_len - uart_tx_buf_offset);
+
+	if (uart_tx_buf_len == uart_tx_buf_offset) {
+		uart_tx_buf_offset = 0;
+		uart_tx_buf_len = 0;
+		uart_tx_buf = uart_tx_buf_bak;
+
+		k_timer_stop(&uart_tx_timer);
+		k_sem_give(&tx_done_sem);
+
+		if (tx_trx_data_cb) {
+			zigbee_schedule_callback(
+				tx_trx_data_cb,
+				SERIAL_SEND_SUCCESS);
+			tx_trx_data_cb = NULL;
+		}
+	} else {
+		k_timer_start(
+			&uart_tx_timer,
+			K_MSEC(CONFIG_ZIGBEE_UART_PARTIAL_TX_TIMEOUT),
+			K_NO_WAIT);
+	}
+}
+
+static void interrupt_handler(const struct device *dev, void *user_data)
+{
+	while (uart_irq_update(dev) && uart_irq_is_pending(dev)) {
+		if (uart_irq_rx_ready(dev)) {
+			handle_rx_ready_evt(dev);
+		}
+
+		if (uart_irq_tx_ready(dev)) {
+			handle_tx_ready_evt(dev);
+		}
+	}
+}
+
+void zb_osif_async_serial_init(void)
+{
+	if (uart_dev != NULL) {
```

Comment:
Is this `if` needed here?

Reply:
If I'm not mistaken - the ZBOSS core may call `zb_osif_serial_init` more than once, so this check is needed.

Reply:
Sorry :) I've badly read the condition.